modules/server_new/src/main/scala/observe/server/odb/OdbProxy.scala

```diff
@@ -0,0 +1,489 @@
+          _.get(obsId)
+            .flatMap(RecordedIds.stepId.get)
+            .toRight(ObserveFailure.Unexpected(s"No current recorded step for obsId [$obsId]"))
+        .rethrow
+
+    private def setCurrentStepId(obsId: Observation.Id, stepId: RecordedStepId): F[Unit] =
+      idTracker.update:
+        ObsRecordedIds.at(obsId).some.andThen(RecordedIds.stepId).replace(stepId.some)
+
+    private val fitsFileExtension                                   = ".fits"
+    @unused private def normalizeFilename(fileName: String): String = if (
+      fileName.endsWith(fitsFileExtension)
+    ) fileName
+    else fileName + fitsFileExtension
+
+    override def datasetStart(
+      obsId:  Observation.Id,
+      fileId: ImageFileId
+    ): F[Boolean] =
+      for {
+        stepId <- getCurrentStepId(obsId)
+        _      <-
+          L.debug(
+            s"Send ODB event datasetStart for obsId: $obsId, stepId: $stepId with fileId: $fileId"
+          )
+        did    <- recordDataset(stepId, fileId)
+        _      <- L.debug(s"Recorded dataset id $did")
+        _      <- AddDatasetEventMutation[F]
+                    .execute(datasetId = did, stg = DatasetStage.StartObserve)
+        _      <- L.debug("ODB event datasetStart sent")
+      } yield true
+
+    private def recordDataset(stepId: RecordedStepId, fileId: ImageFileId): F[DatasetId] =
+      RecordDatasetMutation[F]
+        .execute(stepId.value, normalizeFilename(fileId.value))
+        .map(_.recordDataset.dataset.id)
+
+    override def datasetComplete(
+      datasetId: DatasetId,
+      obsId:     Observation.Id,
+      fileId:    ImageFileId
+    ): F[Boolean] =
+      for {
+        _ <-
+          L.debug(
+            s"Send ODB event datasetComplete for obsId: $obsId datasetId: $datasetId with fileId: $fileId"
+          )
+        // FIXME Data set id is null
+        // _ <- AddDatasetEventMutation[F]
+        //        .execute(datasetId = datasetId, stg = DatasetStage.EndObserve)
+        _ <- L.debug("ODB event datasetComplete sent")
+      } yield true
+
+    override def obsAbort(
+      obsId:  Observation.Id,
+      reason: String
+    ): F[Boolean] =
+      for {
+        visitId <- getCurrentVisitId(obsId)
+        _       <- L.debug(s"Send ODB event observationAbort for obsId: $obsId")
+        _       <- AddSequenceEventMutation[F].execute(vId = visitId, cmd = SequenceCommand.Abort)
+        _       <- L.debug("ODB event observationAbort sent")
+      } yield true
+
+    override def sequenceEnd(obsId: Observation.Id): F[Boolean] =
+      L.debug(s"Skipped sending ODB event sequenceEnd for obsId: $obsId")
+        .as(true)
+
+    override def sequenceStart(
+      obsId:        Observation.Id,
+      instrument:   Instrument,
+      sequenceType: SequenceType,
+      stepCount:    NonNegShort,
+      staticCfg:    StaticConfig
+    ): F[(VisitId, RecordedAtomId)] =
+      for {
+        // TODO Check that there are no current visits or atoms??
```

Comment:
Remember we could be running multiple instruments in parallel

Reply:
Ok. But in that case we use the same visit, atom and step ids? Or do any of them change?

Reply:
I'd say all change, they are per sequence which will be different for each instrument

Reply:
Ok, but they have the same `Observation.Id`? Should we change the key to `(Observation.Id, Instrument)`?

Reply:
They could be different observations